src/java/main/org/apache/zookeeper/server/ZooKeeperServer.java

```diff
@@ -303,15 +305,38 @@ public void loadData() throws IOException, InterruptedException {
 
     public void takeSnapshot(){
```

Comment:
@hanm Thanks for your advice :)

We are using a similar structure as `2.2 Alternative 2 – Learners` in the paper, it is good enough for most of online scenes. I don't think it is worthy to introduce another layer because we give up the write operation to all observers, just use it as a read only view.

For the remote replication, I doubt it also blocked by packet loss rate like Observer. On the other hand, we have an internal project named `zkpipe`, it read Zookeeper snapshot/binlog and send it to a Kafka topic, our client could choose to rebuild the transaction or subscribe the changes. I believe it will better than hack Zookeeper itself. If you have interested, I could push it to github later.